### ISSUE TITLE
Spacer SWE and dev-ops.

### ISSUE THREAD

**Issue by beijbom** [beijbom/coralnet#284]
This ticket is to do SWE and dev-ops for spacer. This includes:

- [x] Create a Dockerfile for better container building process. Support both caffe and pytorch inside the container. 
- [x] Test new container to ensure backwards compatibility.
- [x] Setup Travis CI on GitHub.
- [x] Isolate Caffe such that the rest of repo functions even when Caffe is not available. 
- [x] Make and publish PIP package. This would only suppor the PyTorch. Make sure the PIP package supports both python 2.7 and 3.7.
- [x] Better Modularization of codebase. Abstract away the specifics of the classifiers and feature extractors to allow implementation of new versions.
- [x] Document messages and "API".
- [x] Abstract away storage and support both local and S3.

**Comment by beijbom** [comment]
@StephenChan @qiminchen : FYI: the more I looked at the spacer codebase, the more I realized I need to do some dev-ops and SWE before we add the new feature extractors. This ticket lists the main todos related to this. My goal is that we continue supporting the current feature extractor.

**Comment by StephenChan** [comment]
Okay. If you feel that you're short on time to get all these tasks done, let us know if there's something we can help with.
By SWE you mean software engineering, right?

**Comment by beijbom** [comment]
Hey @StephenChan .  Yeah, Software engineering. I think I got this one. I have done enough thinking on this that it should be relatively smooth sailing. 

The one thing that would help is if I don't have to provide support for both python 2.7 and 3.{6, 7} in the pip package. I haven't looked at it yet, so maybe it is fine, but I do anticipate a bit of headache to get that going. So if you feel ready to move CoralNet to python 3.6 or 3.7, and that you can do it soon, then that would help. If not - and it turns out to be tricky to support both - I'll start with 2.7 only.

**Comment by StephenChan** [comment]
> The one thing that would help is if I don't have to provide support for both python 2.7 and 3.{6, 7} in the pip package. I haven't looked at it yet, so maybe it is fine, but I do anticipate a bit of headache to get that going. So if you feel ready to move CoralNet to python 3.6 or 3.7, and that you can do it soon, then that would help. If not - and it turns out to be tricky to support both - I'll start with 2.7 only.

Well, the thing is, I had the opposite sequence of actions in mind: I figured it'd be a little safer to move to Python 3 once everything is easily testable, including the vision backend. That would mean an issue sequence of #284 -> #264 -> #59. But I guess you're saying #59 -> #284 -> #264 would make your work on this issue easier.

By 'safer', I mean there's a lower chance that some vision backend functionality will break without us noticing when I make code changes for Py3 compatibility. I think I know enough about the Python 2 and 3 differences to make relatively safe changes, but it's easy to miss something.

Still, you know the backend better than I do, so I can leave that risk-tradeoff decision to you. Are you OK with me going ahead with #59 before #264?

**Comment by beijbom** [comment]
Oh, right. I do remember that. And I agree that #284 -> #264 -> #59 would have been the most robust way. But I wonder if we can do it concurrently? Something like this.

1. You cut a "Python 3" feature branch of CoralNet master and make Python 3 migration PRs into that. Test the backend part best you can; it's ok if parts are untested.
2. I simultaneously work on #284 on Spacer.
3. Once both are done, I do #264 and merge that onto the "Python 3" feature branch. This includes adding spacer to requirements and adding tests.

Once step 3 is complete we should have a well-tested system that we can merge to ,aster. The only issue is that we are doing changes without having full test cover. Which I agree is not best practice. But I think it's OK in this case.

**Comment by StephenChan** [comment]
Right, you can work on Spacer concurrently with my Py3 stuff. That branch plan makes sense; let's do that then.

**Comment by beijbom** [comment]
Awesome.

**Comment by kriegman** [comment]
Nice plan!

**Comment by beijbom** [comment]
For visibility, working on this in https://github.com/beijbom/ecs_spacer/pull/5. 

It is more than a PR, really. It is basically a new repo. I'll ping everyone when its time to review.

**Comment by beijbom** [comment]
Ok guys, the new spacer-repo is ready for review. As you will see, there are pretty massive changes. The algorithms are the same, but the whole repo has been reworked as defined in this ticket. Since the changes were so large I didn't put this in a PR, but simply merged it all to master. So please visit https://github.com/beijbom/pyspacer and try installing, testing, and playing around with it.

- The repo is made public and renamed to `pyspacer`.
- CI enabled through travis.
- Test cover is at 100%
- Release through `pip install spacer` at https://pypi.org/project/pyspacer/0.0.2/.
- Support for both caffe and pytorch through Docker with a clean Dockerfile build-file.
- Caffe imports isolated such that the rest of the repo can run without caffe (obviously without support for the caffe-based tasks).
- Data-classes defined for all messages and data types.
- Support for python 3.5 and 3.6 (more details below).
- Storage backends provided for S3, filesystem and memory.
- Scripts  for re-training classifiers on current extracted CoralNet features. These act as regression tests for the train_classifier task. https://github.com/beijbom/pyspacer/blob/master/scripts/classifier_training_regression.py (CC @qiminchen: I checked a few sources, and they all look fine, but please do try it out.)

What happens next? A few things:
- Please review the code carefully, and test all three build concepts (docker, pip install, and git clone) on your local machines. Please let me know if you have thoughts and suggestions. There is a ton of new code so I'm sure there is plenty of room for improvements.
- We have a bit of a problem with scikit-learn. We have been using `0.17.1` to train the current classifiers on CoralNet. This is OK in itself, but python3.7 and 3.8 needs a higher version of scikit-learn, so we'd need to go to `0.22.x` if we want to use those python versions. But here is the problem: scikit-learn doesn't provide backwards compatibility for their serialized (pickled) models (confer: https://github.com/scikit-learn/scikit-learn/issues/10319). Since we want to make spacer a dependency for CoralNet we have three options (1) Stick with scikit-learn 0.17.1 and use python3.5 or 3.6 for CoralNet, (2) Upgrade to scikit-learn `0.22.x` and use the newest python, but then we have to _throw away all old robots and recompute them_. (3) Use option 1 for now and then figure out a way to migrate the model files in the future. 
- Implement the new pytorch feature extractor.

**Comment by kriegman** [comment]
Ouch. Bitten by pickle.
FWIW: We weren't allowed to use pickle at Dropbox. Mostly because of the
security issues of an untrusted pickle file, but also because of these
compatibility issues.

one possibility is if there are alternative serialization methods for the
scikitlearn that are compatible across versions. Read pickle in w/ 0.17.1,
dump in compatible format, ...

D

**Comment by StephenChan** [comment]
Great work, awesome to see all the progress! I'll see how much of the code I can review/test this weekend, but most likely I'll need to finish doing so next week, since I've been putting off primary election voting...

Regarding the scikit-learn problem:

- What about a stance somewhere between (2) and (3)?: Upgrade to scikit-learn 0.22.x, and continue storing the old robots, but hide them for the time being. Thus, we force sources to train a new robot which can be used in 0.22.x, but we don't throw away the old robots yet. If we later figure out a way to migrate the pre-0.22 robots to the new format, we can migrate them and then un-hide them for historical interest.

- Would the situation look any better if I can finish the Python 3 upgrade before we roll out the new spacer system? I was already hoping to focus on Python 3 next, since I can't think of other API specification cleanups to make at the moment, and the other API stuff might not be immediately urgent. Realistically though, I'll need at least 3 solid weeks to complete the Py3 upgrade (including getting the other 3rd party packages up to speed as needed).

**Comment by kriegman** [comment]
FWIW: Logistic regression models are pretty simple with one set of weights
(coefficients) and offset (intercept), and savable as .json without
really needing banfide pickle serialization.
It would be a pain, but you could convert the old models fro pick to json
in old version of scikitlearn and then use json for saving LR models rather
than pickle in the future.

Not sure how much bigger/slower json reading would be or if it matters.

here's a relevant blog post on saving LR as json.
https://medium.com/datadriveninvestor/machine-learning-how-to-save-and-load-scikit-learn-models-d7b99bc32c27

Feel free to ignore suggestion ....

**Comment by qiminchen** [comment]
Great! I will try as much as I can to review and test the new repo this
weekend.

In terms of re-training classifiers on current extracted CoralNet features,
I will definitely try this out to verify the current baseline. And as far
as the newly trained CoralNet evaluation goes, we've seen the accuracy
improvement vary from sources, the comparison and confusion matrices are
available in the research plan
<https://docs.google.com/document/d/1mL2n6KFi2CAcTpMI3OXBnHFIkkfVWBDEtDmGGLRU4hs/edit?usp=sharing>,
these results will be kept updating as soon as I re-train the LR classifier
using the same train/test split (already have some sources re-trained).

The attachment is a sample of the original image where each annotation is
marked with its ground truth label and label predicted by the classifier
(formatted as x: (gt: xxx, est: xxx), correct prediction in yellow and
wrong in white, might change the color later, the classification accuracy
of this particular image is shown in the upper left corner). This might be
a good way to make the coral classification more visible and interpretable
and help us find out whether those misclassified labels are reasonable or
not. (And probably can be added to the backend as an API in the future,
feel free to ignore this)

cheers

 i2227.png
<https://drive.google.com/file/d/1_L0rGA8RUGsBarDh49y2fT8fUu2p-2NM/view?usp=drive_web>

**Comment by beijbom** [comment]
@StephenChan : regarding rolling this out. We can't roll this out until CoralNet is on python3.x, so let's follow the work-order we agreed on above. Once you create a python3.x.base-branch I can start pushing the changes to the vision backend to that, but I'll let you get it to a state where it at least compiles and runs first.

What I _can_ do is deploy this on a new elastic container service & queue such that we can do shakeouts on our dev systems (again, after the 3.x branch is up). I'll do that after we have ironed out the first fews round of bugs, and also agreed on what to do about the scikit-learn. Speaking of which: what python version had you envisioned we go to? The last LRS Ubuntu release (18.04.4) comes with Python 3.6 as standard. So it doesn't seem totally crazy to stay on python 3.6, in which case we can punt the scikit-learn problem down the road a year or two.

@kriegman : I think I understand you, but just to be clear. You are suggesting we extract the learned parameters from the skikit-learn model, store them as JSON and then write the inference step as native python & numpy (independent of scikit-learn)? I like that suggestion a lot. The only complication is that I'm not just doing logistic regression, but I'm training a _calibrated_ logreg. Specifically this one https://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html. Calibration makes inference more complicated, as it involves some sort of weighted sigmoid or logistic function across the classes. I'll take a look at their code to see if I can figure out what they do exactly during inference, and see if I can replicate it. I'll create a new ticket to track that.

@qiminchen . Thanks. Do take your time. We can spend 2-3 weeks on shakeouts and reviews. I'd be neat if you could pip install and call spacer inside your code-base so that we know for sure we run the same LR training code.

**Comment by kriegman** [comment]
I'm thinking of something simpler.

I haven't looked at your code, but I presume that the issue is simply the
pickling, and the same weights could be used for either version of
sckitlearn. So, my suggestion is to replace pickle with you own serializing
that saves to json. A 2 minute look at the documentation shows
get_params, set_params.

I know, easier said than done...

**Comment by beijbom** [comment]
Ok, yeah, I was considering that option as well. The problem is that the
object definition itself that has changed, not the pickle protocol, so it
may not even have the same parameter set. Or indeed implement the same
algorithm. :/

**Comment by kriegman** [comment]
Ouch.

**Comment by StephenChan** [comment]
> regarding rolling this out. We can't roll this out until CoralNet is on python3.x, so let's follow the work-order we agreed on above. Once you create a python3.x.base-branch I can start pushing the changes to the vision backend to that, but I'll let you get it to a state where it at least compiles and runs first.

Ah, okay, I get it now. Guess I wasn't paying enough attention there.

I'll start by adding a few more unit tests, and we can merge those in separately first. Then once I'm really working on the Python 3.x compatibility, I'll work in that branch as agreed previously.

> What I _can_ do is deploy this on a new elastic container service & queue such that we can do shakeouts on our dev systems (again, after the 3.x branch is up). I'll do that after we have ironed out the first fews round of bugs, and also agreed on what to do about the scikit-learn.

That sounds like a good idea.

> Speaking of which: what python version had you envisioned we go to? The last LRS Ubuntu release (18.04.4) comes with Python 3.6 as standard. So it doesn't seem totally crazy to stay on python 3.6, in which case we can punt the scikit-learn problem down the road a year or two.

Hmm... that's a good point. In terms of Python features, I think 3.6 will already feel pretty great in comparison to 2.7. So my only concern would be the Python 3.6 end of life date, which appears to be the start of 2022 (https://www.python.org/dev/peps/pep-0494/). Presumably, most third-party packages would support 3.6 until then as well. So yes, unless any packages jump the gun on dropping 3.6, I think we'd indeed have a year or two to work out the serialization issue. That plan would be fine with me.

**Comment by StephenChan** [comment]
I saw that https://github.com/beijbom/pyspacer/pull/5/ requires boto 2.49.0 and numpy 1.17.0. If coralnet requires spacer, then we'll need to sync coralnet's boto and numpy versions accordingly, meaning we should check that coralnet works fine with those versions. So that's another thing to upgrade on top of Py3.

**Comment by beijbom** [comment]
Numpy 1.17.5 actually. Yeah. But I’m not so worried about that one, it is a
very recent release. The boto 2.49.0 release is the last one on boto2, so
should be a good  stable one. You could use boto3 for CoralNet if we want
since it’s a different package. I just stayed on boto2 since I was familiar
with the interface.

**Comment by StephenChan** [comment]
Yeah, I just realized https://github.com/beijbom/pyspacer/pull/5 didn't have the most up to date requirements. I'll refer to the master branch from now on.

**Comment by StephenChan** [comment]
@beijbom I'm wondering, did you get some particular error when trying to install scikit-learn 0.17.1 on Python 3.7/3.8, or did you just read the compatibility notes? I'm noticing that the Python versions for packages might be more lenient than I thought.

For example, I tried installing Django 1.9 on Python 3.7 and it seemed to work, despite the release notes stating that Django 1.9 would work with Python 2.7, 3.4 or 3.5.

I didn't try with scikit-learn itself yet, since I seem to have some issue installing it in general right now (seems Windows related).

**Comment by beijbom** [comment]
Yeah. I got errors. Both in the Travis CI and on my dev system. I don’t
recall the exact error, but it was resolved when moving to 0.22.

**Comment by StephenChan** [comment]
Okay, just figured I'd double check.

I also triple checked by getting further with Windows installations myself. Results:

- Python 3.6 + scikit-learn 0.17.1: works
- Python 3.7 + scikit-learn 0.17.1: errors such as `sklearn\svm\libsvm.c(10699): error C2039: 'exc_type': is not a member of '_ts'`
- Python 3.7 + scikit-learn 0.18.2: errors like the above
- Python 3.7 + scikit-learn 0.19.2: works
- Python 3.7 + scikit-learn 0.20.4: works
- Python 3.7 + scikit-learn 0.21.3: works, after doing `pip install Cython` first
- Python 3.7 + scikit-learn 0.22.1: works, after doing `pip install Cython` first

I used the `--no-binary :all:` option to ensure all of those were source installations, not binary.

So, one more check:

- Do you think going from sklearn 0.17.1 to 0.19.2 would have the same model portability problem?
- If not, does Py 3.7 + sklearn 0.19.2 install for you?

**Comment by beijbom** [comment]
On Tue, Mar 3, 2020 at 01:47 StephenChan <notifications@github.com> wrote:

> Okay, just figured I'd double check.
>
great.

> I also triple checked by getting further with Windows installations
> myself. Results:
>
>    - Python 3.6 + scikit-learn 0.17.1: works
>    - Python 3.7 + scikit-learn 0.17.1: errors such as sklearn\svm\libsvm.c(10699):
>    error C2039: 'exc_type': is not a member of '_ts'
>    - Python 3.7 + scikit-learn 0.18.2: errors like the above
>    - Python 3.7 + scikit-learn 0.19.2: works
>    - Python 3.7 + scikit-learn 0.20.4: works
>    - Python 3.7 + scikit-learn 0.21.3: works, after doing pip install
>    Cython first
>    - Python 3.7 + scikit-learn 0.22.1: works, after doing pip install
>    Cython first
>
> I used the --no-binary :all: option to ensure all of those were source
> installations, not binary.
>
> So, one more check:
>
>    - Do you think going from sklearn 0.17.1 to 0.19.2 would have the same
>    model portability problem?
>
> Did the tests run? If they did we should be good to go.


>    -
>    - If not, does Py 3.7 + sklearn 0.19.2 install for you?
>
> You can create a branch, change the .jenkins (just add a new line for 3.7)
and the requirements files to see if it works there. If it does we are good
to go.

Thanks for double checking all of this.

Btw., IMHO ,if to go beyond 3.6 we should probably to go 3.8.

**Comment by beijbom** [comment]
On Tue, Mar 3, 2020 at 06:52 Oscar Beijbom <oscar.beijbom@gmail.com> wrote:

>
>
> On Tue, Mar 3, 2020 at 01:47 StephenChan <notifications@github.com> wrote:
>
>> Okay, just figured I'd double check.
>>
> great.
>
>> I also triple checked by getting further with Windows installations
>> myself. Results:
>>
>>    - Python 3.6 + scikit-learn 0.17.1: works
>>    - Python 3.7 + scikit-learn 0.17.1: errors such as sklearn\svm\libsvm.c(10699):
>>    error C2039: 'exc_type': is not a member of '_ts'
>>    - Python 3.7 + scikit-learn 0.18.2: errors like the above
>>    - Python 3.7 + scikit-learn 0.19.2: works
>>    - Python 3.7 + scikit-learn 0.20.4: works
>>    - Python 3.7 + scikit-learn 0.21.3: works, after doing pip install
>>    Cython first
>>    - Python 3.7 + scikit-learn 0.22.1: works, after doing pip install
>>    Cython first
>>
>> I used the --no-binary :all: option to ensure all of those were source
>> installations, not binary.
>>
>> So, one more check:
>>
>>    - Do you think going from sklearn 0.17.1 to 0.19.2 would have the
>>    same model portability problem?
>>
>> Did the tests run? If they did we should be good to go.
>
>
>>    -
>>    - If not, does Py 3.7 + sklearn 0.19.2 install for you?
>>
>> You can create a branch, change the .jenkins (just add a new line for
> 3.7) and the requirements files to see if it works there. If it does we are
> good to go.
>
*and push to GitHub obviously.

**Comment by StephenChan** [comment]
Ah, sorry, when I said 'works' I just meant that I was able to install the packages. I had not tried running the spacer tests.

Unfortunately, after trying the spacer tests (changing requirements/setup files as needed), it looks like scikit-learn 0.19.2 won't load legacy models:  

```
======================================================================
ERROR: test_load_legacy_model (test_storage.TestLocalStorage)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\<...>\virtualenv_pyspacer_py37_sklearn0192\Lib\site-packages\spac
er\tests\test_storage.py", line 142, in test_load_legacy_model
    config.LOCAL_FIXTURE_DIR, 'legacy.model'))
  File "D:\<...>\virtualenv_pyspacer_py37_sklearn0192\lib\site-packages\spac
er\storage.py", line 106, in load_classifier
    return pickle.load(f, encoding='latin1')
ModuleNotFoundError: No module named 'copy_reg\r'
```

I've also tried several combinations including Python 3.8 now. The overall conclusion is that we do have to stay on Python 3.6 to load legacy models. (Again, I'm fine with this, but wanted to make sure.) scikit-learn 0.18.2 *might* be okay, if we're lucky, but that version's still Py3.6 only so it wouldn't matter much. Here are all of my results:

### Python 3.6

- scikit-learn 0.17.1: Installs. Spacer tests on legacy models pass.
  - Overall test results: no fails, no errors.
- scikit-learn 0.18.2: Installs. Spacer tests on legacy models pass, but get warnings.
  - Overall test results: no fails, 1 error. The error is on TestDeploy.test_deploy_simple().
  - The warnings are of the form `UserWarning: Trying to unpickle estimator SGDClassifier from version pre-0.18 when using version 0.18.2. This might lead to breaking code or invalid results. Use at your own risk.`

### Python 3.7

- scikit-learn 0.17.1, 0.18.2: Can't install.
  - Gets errors of the form `sklearn\svm\libsvm.c(10699): error C2039: 'exc_type': is not a member of '_ts'`
- scikit-learn 0.19.2, 0.20.4, 0.21.3, 0.22.1: Installs. Spacer tests on legacy models get an error.
  - The legacy model error is of the form `ModuleNotFoundError: No module named 'copy_reg\r'`. Overall test results: 5 fails, 2 errors.

### Python 3.8

- scikit-learn 0.17.1, 0.18.2, 0.19.2, 0.21.3, 0.22.1: Same results as Python 3.7.
- scikit-learn 0.20.4: Installs. But only 33 Spacer tests run (instead of 93) and there are 7 errors of the form `TypeError: an integer is required (got type bytes)`.

Not sure what's up with the Py3.8 + 0.20.4 case, but I guess it's not likely to matter.

On the bright side, I've confirmed that pyspacer can be installed on Windows. I only tried the pip install method so far. I should note that I had to copy the `tests/fixtures` subdirectory manually into my installation, because pip-install didn't include that directory automatically. I think pip install excludes non-.py files by default, in which case we might fix that by using distutils' `data_files` option: https://stackoverflow.com/questions/12191339/

**Comment by beijbom** [comment]
Thanks for double checking this @StephenChan. For now, let's assume we stay on python 3.6 and those other libraries I found. If I can figure out a better way to do the serialization we can revisit this again. 

Re. missing fixture files in the pip package. I actually think that is expected behavior. Indeed, I was meaning to exclude the test code itself in the package. One does usually not develop and run tests on a pip-installed package, right?  One would just assume it is packaged from a stable branch/build. So for running tests I'd say clone and use a virtualenv. or the docker build. Correct me if I'm wrong here, but at least I have never thought of trying to run tests on a pip-installed package.

Have you tried using Docker yet? If you haven't-- give it a go. The tooling is pretty amazing actually. The first time you build the image it will take a while, but subsequent builds are super fast.

**Comment by StephenChan** [comment]
> Re. missing fixture files in the pip package. I actually think that is expected behavior. Indeed, I was meaning to exclude the test code itself in the package. One does usually not develop and run tests on a pip-installed package, right? One would just assume it is packaged from a stable branch/build. So for running tests I'd say clone and use a virtualenv. or the docker build. Correct me if I'm wrong here, but at least I have never thought of trying to run tests on a pip-installed package.

Oh yeah, you're right. I forgot about that.

> Have you tried using Docker yet? If you haven't-- give it a go. The tooling is pretty amazing actually. The first time you build the image it will take a while, but subsequent builds are super fast.

Haven't tried it yet, that'll be the next step. Sounds pretty promising.

**Comment by StephenChan** [comment]
I tried Docker. Apparently Docker is a pain to install on Windows 10 Home edition, so I tried it out in Ubuntu server. Got all the tests passing. Notes:

- Installed `docker.io` from apt on Ubuntu 18.04.
- Cloned the pyspacer repo 3-4 hours ago, so it's the current master branch.
- At first I didn't realize that `SPACER_LOCAL_MODEL_PATH` in `secrets.json` had to be `/workspace/models` if using Docker. I initially used the path of the folder I created in Ubuntu, which gave me some FileNotFoundErrors when running tests. @beijbom Maybe the README's installation instructions could clarify this part better?
- The initial Docker build took 10 minutes, but the next one (after editing secrets.json) took 3 seconds.

Other than the model path thing, it was quite smooth to set up, so nice job on that!

**Comment by beijbom** [comment]
Thanks for trying it @StephenChan , I'm glad it worked!

Re that path, I have already fixed in the last open PR by setting the env. variable inside the docker. That has priority over the content of `secrets.json`. (Line 80 of Dockerfile in https://github.com/beijbom/pyspacer/pull/10/files)

**Comment by StephenChan** [comment]
Oh, good to know you already fixed that! I'll keep an eye on your PR updates and at least glance over them. But let me know if there's any part you particularly want reviewed.

**Comment by StephenChan** [comment]
This can be closed, right?

**Comment by beijbom** [comment]
Yeah, thanks.
